nixos/modules/system/boot/stage-2-init.sh

```diff
@@ -61,25 +61,42 @@ else
 fi
 
 
-# Make /nix/store a read-only bind mount to enforce immutability of
-# the Nix store.  Note that we can't use "chown root:nixbld" here
+# Give /nix/store the defined mount options.
+# Typically, this should be:
+# - 'ro' to enforce immutability of the Nix store
+# - 'nosuid' to enforce no suid binaries make it into the store and get executed by accident.
+#   suid-binaries should only exist in /run/wrappers.
+#   If an attacker can make the nix builder produce suid binaries in the store, those should be useless.
+#   Another example is tampering with the store from an outside system.
+# - 'nodev' to enforce no device files in the store
+# Note that we can't use "chown root:nixbld" here
 # because users/groups might not exist yet.
+
 # Silence chown/chmod to fail gracefully on a readonly filesystem
 # like squashfs.
 chown -f 0:30000 /nix/store
 chmod -f 1775 /nix/store
-if [ -n "@readOnlyNixStore@" ]; then
+
+missing_opts=() # stores the missing mount options that still need to be applied to the nix store
+current_opts="$(findmnt --direction backward --first-only --noheadings --output OPTIONS /nix/store)"
+for mount_opt in @nixStoreMountOpts@ ; do
     # #375257: Ensure that we pick the "top" (i.e. last) mount so we don't get a false positive for a lower mount.
-    if ! [[ "$(findmnt --direction backward --first-only --noheadings --output OPTIONS /nix/store)" =~ (^|,)ro(,|$) ]]; then
-        if [ -z "$container" ]; then
-            mount --bind /nix/store /nix/store
-        else
-            mount --rbind /nix/store /nix/store
-        fi
-        mount -o remount,ro,bind /nix/store
+    if ! [[ "$current_opts" =~ (^|,)"$mount_opt"(,|$) ]]; then
```

Comment:
This probably doesn't need a comment to be added to the code, but I would appreciate if you'd leave a github comment explaining this syntax.

Reply:
In #375257, we changed the regex to detect mount options to the above because the previous one was `ro(,|$)`. This was a false positive on things like `errors=remount-ro,foo` and resulted in the Nix store not being remounted ro. This regex instead matches any of `ro`, `ro,foo`, `foo,ro`, or `foo,ro,bar`, except with any of the given mount options instead of just ro.

Reply:
Eh, doesn't hurt to have a comment in the code, added.